hw/drivers/ram/spiram/src/spiram.c

```diff
@@ -0,0 +1,286 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <string.h>
+#include <assert.h>
+
+#include "os/mynewt.h"
+#include <bsp/bsp.h>
+#include <hal/hal_spi.h>
+#include <hal/hal_gpio.h>
+#include <spiram/spiram.h>
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+#include <bus/drivers/spi_common.h>
+#endif
+
+static inline void
+spiram_lock(struct spiram_dev *dev)
+{
+#if MYNEWT_VAL(OS_SCHEDULING)
+    os_mutex_pend(&dev->lock, OS_TIMEOUT_NEVER);
+#endif
+}
+
+static inline void
+spiram_unlock(struct spiram_dev *dev)
+{
+#if MYNEWT_VAL(OS_SCHEDULING)
+    os_mutex_release(&dev->lock);
+#endif
+}
+
+void
+spiram_cs_activate(struct spiram_dev *dev)
+{
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+    struct bus_spi_node *node = (struct bus_spi_node *)&dev->dev;
+    hal_gpio_write(node->pin_cs, 0);
+#else
+    hal_gpio_write(dev->ss_pin, 0);
+#endif
+}
+
+void
+spiram_cs_deactivate(struct spiram_dev *dev)
+{
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+    struct bus_spi_node *node = &dev->dev;
+    hal_gpio_write(node->pin_cs, 1);
+#else
+    hal_gpio_write(dev->ss_pin, 1);
+#endif
+}
+
+int
+spiram_write_enable(struct spiram_dev *dev)
+{
+    int rc = 0;
+    spiram_lock(dev);
+
+    if (dev->characteristics->write_enable_cmd) {
+#if MYNEWT_VAL(BUS_DRIVER_PRESENT)
+        rc = bus_node_simple_write((struct os_dev *)&dev->dev,
+                              &dev->characteristics->write_enable_cmd, 1);
+#else
+        spiram_cs_activate(dev);
+
+        rc = hal_spi_tx_val(dev->spi_num, dev->characteristics->write_enable_cmd);
+
+        spiram_cs_deactivate(dev);
+#endif
+    }
+
+    spiram_unlock(dev);
+
+    return rc;
+}
+
+int
+spiram_read(struct spiram_dev *dev, uint32_t addr, void *buf, uint32_t size)
+{
+    uint8_t cmd[5] = { SPIRAM_READ };
+    int i;
```

Comment:
Why is this an int ? A uint8_t is enough.

Reply:
int is a native type and it's usage usually produces more optimal code.
In this case for ARM build both debug and optimized profiles result in shorter binary when int is used instead of uint8_t.